include/zephyr/drivers/uart.h

```diff
@@ -1655,6 +1655,28 @@ static inline int z_impl_uart_drv_cmd(const struct device *dev, uint32_t cmd,
 #endif
 }
 
+#define UART_DT_INST_NAME_PREFIX UTIL_CAT(uart_, DT_DRV_COMPAT)
+
+#define UART_DT_INST_API_NAME(n) UTIL_CAT(UART_DT_INST_NAME_PREFIX, _api)
+
+#define UART_DT_CONFIG_NAME(node_id) UTIL_CAT(Z_DEVICE_DT_DEV_NAME(node_id), _config)
+#define UART_DT_DATA_NAME(node_id) UTIL_CAT(Z_DEVICE_DT_DEV_NAME(node_id), _data)
+
+#define UART_DT_INST_CONFIG_NAME(inst) UART_DT_CONFIG_NAME(DT_DRV_INST(inst))
+#define UART_DT_INST_DATA_NAME(inst) UART_DT_DATA_NAME(DT_DRV_INST(inst))
+
+#define UART_DT_INST_API_DECLARE \
+	static const struct uart_driver_api UART_DT_INST_API_NAME
+
+#define UART_DT_INST_DEFINE(n, init_fn) 			\
+	DEVICE_DT_INST_DEFINE(n, init_fn, PM_DEVICE_DT_INST_GET(n),	\
+			    &UART_DT_INST_DATA_NAME(n),			\
+			    &UART_DT_INST_CONFIG_NAME(n),		\
+			    PRE_KERNEL_1,				\
+			    CONFIG_SERIAL_INIT_PRIORITY,		\
```

Comment:
this is not always true, certain devices have their own custom init levels/priorities on a subsystem. One example is GPIO expanders, tipically connected to a bus, so they need a custom priority to make sure they are initialized after bus. I think init priorities need to be removed from DT devices before being able to do this.

Reply:
I'm aware.  Figured these macros will differ a little based on subsystem.  For things like UART where its common we do this.  We can figure out if we expose a macro that lets setting the priority/level for those cases that need it.

Reply:
If more customization is desired, an include with defaults being defined if values are not already defined could be provided... e.g.

uart_macros.h

``` c
#ifndef UART_INIT_PRIO
#define UART_INIT_PRIO CONFIG_SERIAL_INIT_PRIORITY
#endif

#define UART_DT_INST_DEFINE(n, init_fn) \
     ....
```

This is sort of the idea I had in my alternative to defining a new I2C_DEVICE_DT_DEFINE macro for i2c when I did that for stats. Maybe some useful ideas in this dead end PR still:

https://github.com/zephyrproject-rtos/zephyr/pull/42880